common/ostream.h

```diff
@@ -62,6 +62,35 @@ void PrintTo(T* p, std::ostream* out) {
   }
 }
 
+// True if T can be used with `<<` and an `llvm::raw_ostream`.
+template <typename T>
+static constexpr bool HasRawOstream = Requires<const T, llvm::raw_ostream>(
+    [](auto&& t, auto&& out) -> decltype(out << t) {});
+
+// Provide an `operator<<` which detects types with `raw_ostream` overloads and
+// uses that to map to a `std::ostream` overload for enum types. Other types
+// should simply provide a `Print` method which will give them all of these
+// while also being more usable from a debugger. But enums cannot add this so we
+// let them define a single operator and connect to it here.
+//
+// This overload is disabled for non-enum types, or when there is no
+// `raw_ostream` streaming support. And to make this overload be unusually low
+// priority, it is designed to take even the `std::ostream` parameter as a
+// template, and SFINAE disable itself unless that template parameter matches
+// `std::ostream`. This ensures that an *explicit* operator will be preferred
+// when provided.
```

Comment:
I'm confused by this, and what it prints for an enum. Can you please add a unit test to show the result? Also, what's the specific use-case that's making you add it? Is this for gtest, and should that be mentioned (and maybe even testonly under //testing rather than throughout code)?

Note in part, I thought we'd been using the EnumBase to provide reliable printing.

Reply:
Okay, if as TL you don't think this needs to be tested I'll move on.